src/net/discovery/conf.rs

```diff
@@ -0,0 +1,118 @@
+struct Conf {
+    peers: [peer::Config; 3],
+}
+
+impl Conf {
+    /// Reads config from `file_location`. Expects file to be json format
+    /// # Errors
+    /// if the file does not exist, or is in an invalid format
+    #[allow(dead_code)] // TODO: will use in upcoming PR
+    pub fn new(file_location: &str) -> Result<Self, Error> {
+        use config::{Config, File, FileFormat};
+        // use std::{fs::File, io::BufReader};
+
+        // let contents = File::open(file_location)?;
+        // let buffered = BufReader::new(contents);
+        Ok(Config::builder()
+            .add_source(File::new(file_location, FileFormat::Toml))
+            .build()?
+            .try_deserialize()?)
+    }
+}
+
+impl PeerDiscovery for Conf {
+    fn peers(&self) -> [peer::Config; 3] {
+        self.peers.clone()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::net::discovery::peer::PublicKey;
+    use hyper::Uri;
+    use rand::{
+        distributions::Alphanumeric,
+        {thread_rng, Rng},
+    };
+    use std::fs::File;
+    use std::io::Write;
+
+    const H1_PUBLIC_KEY: [u8; 32] = [
+        19, 204, 244, 38, 60, 236, 188, 48, 245, 14, 106, 139, 156, 135, 67, 148, 61, 221, 230, 32,
+        121, 88, 11, 192, 185, 1, 155, 5, 186, 143, 233, 36,
+    ];
+    const H2_PUBLIC_KEY: [u8; 32] = [
+        146, 91, 249, 130, 67, 207, 112, 183, 41, 222, 29, 117, 191, 79, 230, 190, 152, 169, 134,
+        96, 131, 49, 219, 99, 144, 43, 130, 161, 105, 29, 193, 59,
+    ];
+    const H3_PUBLIC_KEY: [u8; 32] = [
+        18, 192, 152, 129, 161, 199, 169, 45, 28, 112, 217, 234, 97, 157, 122, 224, 104, 75, 156,
+        180, 94, 204, 32, 123, 152, 239, 48, 236, 33, 96, 160, 116,
+    ];
+    const H1_URI: &str = "http://localhost:3000";
+    const H2_URI: &str = "http://localhost:3001";
+    const H3_URI: &str = "http://localhost:3002";
+    const EXAMPLE_CONFIG: &str = r#"
+# H1
+[[peers]]
+    [peers.http]
+        origin = "http://localhost:3000"
+        public_key = "13ccf4263cecbc30f50e6a8b9c8743943ddde62079580bc0b9019b05ba8fe924"
+    [peers.prss]
+        public_key = "13ccf4263cecbc30f50e6a8b9c8743943ddde62079580bc0b9019b05ba8fe924"
+
+# H2
+[[peers]]   
+    [peers.http]
+        origin = "http://localhost:3001"
+        public_key = "925bf98243cf70b729de1d75bf4fe6be98a986608331db63902b82a1691dc13b"
+    [peers.prss]
+        public_key = "925bf98243cf70b729de1d75bf4fe6be98a986608331db63902b82a1691dc13b"
+
+# H3
+[[peers]]
+    [peers.http]
+        origin = "http://localhost:3002"
+        public_key = "12c09881a1c7a92d1c70d9ea619d7ae0684b9cb45ecc207b98ef30ec2160a074"
+    [peers.prss]
+        public_key = "12c09881a1c7a92d1c70d9ea619d7ae0684b9cb45ecc207b98ef30ec2160a074"
+"#;
+
+    fn origin_to_string(origin: &peer::Origin) -> String {
+        let uri = Uri::from(origin.clone());
+        format!("{}://{}", uri.scheme().unwrap(), uri.authority().unwrap())
+    }
+
+    #[test]
+    fn config_from_file() {
```

Comment:
since there's no way to feed `config` crate a string literal, the only test here will be read from file

Reply:
ah, yeah that does indeed work. thanks